### PR TITLE
kubenet: replace gateway with cni result

### LINKED ISSUE #84541: Kubenet not working with pod CIDRs from node spec in 1.16
**What happened**:

I have a node with a CIDR allocated using k8s ipam:
```
$ kubectl get nodes -o yaml

<...>

spec:
    podCIDR: 10.47.128.0/24
    podCIDRs:
    - 10.47.128.0/24

<...>
```
but kubenet cannot allocate addresses for pods. Here is how kubenet configures host-local ipam:

```
I1029 22:10:21.832778    7477 kubenet_linux.go:284] CNI network config set to {
  "cniVersion": "0.1.0",
  "name": "kubenet",
  "type": "bridge",
  "bridge": "cbr0",
  "mtu": 1500,
  "addIf": "eth0",
  "isGateway": true,
  "ipMasq": false,
  "hairpinMode": false,
  "ipam": {
    "type": "host-local",
    "ranges": [
[{
"subnet": "10.47.128.1/24",
"gateway": "10.47.128.1"
}]],
    "routes": [{"dst": "0.0.0.0/0"}]
  }
}
```

note, that `subnet` field is 10.47.128.1/24, not 10.47.128.0/24

The error is like that:
```
E1029 22:59:09.005667   18180 remote_runtime.go:105] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = failed to set up sandbox container "95c4515535cff7239dc99e5880689a5537aaf3c5ea146264d4a57ef46e2ce592" network for pod "kube-dns-autoscaler-7b6ddcc6f9-ltc8j": networkPlugin kubenet failed to set up pod "kube-dns-autoscaler-7b6ddcc6f9-ltc8j_kube-system" network: error adding container to network: invalid range set 0: Network has host bits set. For a subnet mask of length 24 the network address is 10.47.128.0
```

**What you expected to happen**:

Pods are expected to run in the node

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

on 1.15 [the behavior is different](https://github.com/kubernetes/kubernetes/blob/v1.15.5/pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go#L248), 10.47.128.0/24 would be configured by kubenet for `CNI`. The [new behaviour](https://github.com/kubernetes/kubernetes/blob/v1.16.2/pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go#L276) was introduced with [dualstack support](https://github.com/kubernetes/kubernetes/commit/dba434c4ba0bf39275f25b3833fcc5f6eefa9379). It looks like this behaviour may have been introduced by a mistake, but I'm not sure.

CNI plugin [check](https://github.com/containernetworking/plugins/blob/v0.8.2/plugins/ipam/host-local/backend/allocator/range.go#L46) fails for such subnets, since it treats CIDR parsing to `IPNet` [a bit different](https://github.com/containernetworking/cni/blob/master/pkg/types/types.go#L36) from the standard `net` package.

**Environment**:
- Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:09:08Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
yandex cloud (cloud.yandex.com)
- OS (e.g: `cat /etc/os-release`):
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

- Kernel (e.g. `uname -a`):
Linux amcb4hb9dd7b6djtgjro-ovav 4.15.0-29-generic #31-Ubuntu SMP Tue Jul 17 15:39:52 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
n/a
- Network plugin and version (if this is a network-related bug):
kubenet, CNI plugins 0.8.2
- Others:
```
KUBELET_OPTS="--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubeconfig.conf --cert-dir=/var/lib/kubelet/pki/ --cloud-provider=external --cni-bin-dir=/home/kubernetes/cni/bin --config=/home/kubernetes/kubelet-config.yaml --experimental-kernel-memcg-notification=true --feature-gates=CSINodeInfo=true,CSIDriverRegistry=true --kubeconfig=/etc/kubernetes/kubelet-kubeconfig.conf --network-plugin=kubenet --node-labels=yandex.cloud/node-group-id=c49s9oshfcqoovjp6m04,yandex.cloud/preemptible=false --node-status-max-images=25 --pod-infra-container-image=k8s.gcr.io/pause:3.1 --resolv-conf=/run/systemd/resolve/resolv.conf --v=5"
```

### PR DESCRIPTION
**Pull request by chendotjs**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Remove the tricky method that `kubenet` uses to calculate gateway from pod cidr. Instead，`kubenet` will fetch gateway from cni result since `host-local`has the ability to conclude an appropriate gateway for a subnet as @mars1024 methioned in #84550.

**Which issue(s) this PR fixes**:

Fixes #84541

**Special notes for your reviewer**:

Thanks for review. 

**Does this PR introduce a user-facing change?**:

```release-note
fix a regression in kubenet that prevent pods to obtain ip addresses
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by chendotjs** [comment]
/priority important-soon
/cc @khenidak @mars1024 @aojea 

I'll keep on adding tests if removing gateway generation is accepted

**Comment by chendotjs** [comment]
ping @mars1024

**Comment by chendotjs** [comment]
ping @mars1024 
ptal

**Review by mars1024 — COMMENTED** [review]
Sorry for the late reply, I think this almost LGTM except some tests :)

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go [hunk 1]
```diff
@@ -336,6 +332,9 @@ func (plugin *kubenetNetworkPlugin) Capabilities() utilsets.Int {
 // setup sets up networking through CNI using the given ns/name and sandbox ID.
 func (plugin *kubenetNetworkPlugin) setup(namespace string, name string, id kubecontainer.ContainerID, annotations map[string]string) error {
 	var ipv4, ipv6 net.IP
+	var podGateways []net.IP
+	var podCIDRs []net.IPNet
```
Comment (aojea) [inline pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go]: nit, we are going to have a method var named `plugin.podCIDR`s and this local var `podCIDRs` , can it be misleading?

**Comment by aojea** [comment]
@chendotjs those tests failures are because you have to update  kubelet/dockershim/network/kubenet/kubenet_linux_test.go

This is indeed better than only fixing the bug

**Comment by chendotjs** [comment]
Yes, the tests will surely fail since I remove `plugin.podGateways`.
I am waiting to see if removing gateway-generation logic in kubenet is accepted by folks here :)

/assign @dcbw

**Comment by chendotjs** [comment]
/assign @thockin 
ptal if you get a chance

**Comment by chendotjs** [comment]
@mars1024 @aojea 
tests have been updated. would you please take another look?  thanks for review :)

**Comment by aojea** [comment]
I suggest to squash the commits before merge, anyway it LGTM, but defer to @mars1024 the final lgtm

**Comment by chendotjs** [comment]
@aojea  thanks for your relpy, the commits are squashed.

**Comment by squeed** [comment]
I think this is missing something: you still need to keep `plugin.podGateways`, because that's where we gather state (though it's still wrong, because it doesn't survive reboots... sigh.

**Comment by chendotjs** [comment]
I'm afraid I didn't catch your meaning of gathering state from `plugin.podGateways`, could you explain a little bit more? @squeed Thanks in advance.
PS: PR #84550 still keeps `plugin.podGateways`.

**Comment by dcbw** [comment]
> I think this is missing something: you still need to keep `plugin.podGateways`, because that's where we gather state (though it's still wrong, because it doesn't survive reboots... sigh.

@squeed I don't know what state gets kept in there. podGateways doesn't seem to be used for anything other than creating the host-local ranges, which doesn't need state to be kept.

**Comment by baranov1ch** [comment]
shouldn't it be cherry-picked to 1.16 and 1.17 ?

**Comment by aojea** [comment]
@chendotjs do you mind adding a release note?
I've added this to the cherry picks
>
>```release-note
>fix a regression in kubenet  that  prevent pods to obtain ip addresses
>```

**Comment by chendotjs** [comment]
@aojea thanks for reminding，release note updated 😃

**Pull request merged**